Add python-releases.toml

AA-Turner opened the pull request:
Inspired by #4314, this PR adds a transcription of every Python release since version 1.6 into a single TOML document, `python-releases.toml`. This is intended to serve as a single, centralised, machine-readable record of Python's release history (and future).

From this, we automatically generate a `release-cycle.json` file as part of the build process, to be published on `peps.python.org`. This replaces the [version in the devguide](https://github.com/python/devguide/blob/HEAD/include/release-cycle.json).

The TOML document is used to (re-)renerate the release schedules contained in release PEPs, initially starting with those for Python 3.8 to 3.14. The authoritative record and history remains the release PEP.

Some releases may need optional annotations or notes, which I have filled in for Python 3.8 and 3.9, but not yet back-filled.

Open questions:

* Any better ideas for a filename than `python-releases.toml`?
* Should the file live at the top level, or in the a subdirectory (as at present)?
* Any better ideas for the metadata field names? I'm not a great fan of `start-of-development` and `end-of-bugfix`, as all the others can be said aloud as "The {first release / feature freeze / end of life} is/was on {date}".
* Are the release managers for Pythons 1.6-2.2 correct?

A

hugovk reviewed (commented):
Partial review: 

> From this, we automatically generate a `release-cycle.json` file as part of the build process, to be published on `peps.python.org`. This replaces the [version in the devguide](https://github.com/python/devguide/blob/HEAD/include/release-cycle.json?rgh-link-date=2025-03-29T18%3A53%3A42Z).

Can we make all the data in the TOML also available in the JSON?

JSON is more universally supported, and we should make this data as widely usable as possible.

We don't necessarily need the JSON here to be an exact copy of the old devguide one; as long as the devguide can read the data it needs from this JSON to construct the diagram and tables.


> Open questions:
> 
> * Any better ideas for a filename than `python-releases.toml`?

Fine by me.

> * Any better ideas for the metadata field names? I'm not a great fan of `start-of-development` and `end-of-bugfix`, as all the others can be said aloud as "The {first release / feature freeze / end of life} is/was on {date}".

We could use the same names as the https://endoflife.date/ API.

The alpha https://endoflife.date/api/python.json currently uses:

* `releaseDate` - initial release
* `support` - end of bugfix
* `eol` - end of security support / life

Or the WIP v1 API (https://github.com/endoflife-date/endoflife.date/pull/2080), https://deploy-preview-2080--endoflife-date.netlify.app/api/v1/products/python uses:

* `date` - initial release
* `eoasFrom` - end of bugfix (end of active support)
* `eolFrom` - end of security support / life

AA-Turner commented on pep_sphinx_extensions/pep_zero_generator/pep_index_generator.py:
```diff
@@ -19,13 +19,19 @@
 from pep_sphinx_extensions.pep_zero_generator import subindices
 from pep_sphinx_extensions.pep_zero_generator import writer
 from pep_sphinx_extensions.pep_zero_generator.constants import SUBINDICES_BY_TOPIC
+if sys.version_info >= (3, 11):
+    from release_engineering.generate_release_cycle import create_release_cycle
+else:
+    def create_release_cycle():
+        return ''
```
Done (it requires tomllib, but we can change the minimum supported version in a different PR)

AA-Turner commented on release_engineering/LICENCE.rst:
```diff
@@ -0,0 +1,2 @@
+This files in this directory are placed in the public domain or under the
```
Oops, good spot.

AA-Turner commented on release_engineering/update_release_schedules.py:
```diff
@@ -0,0 +1,167 @@
+
+TODAY = dt.date.today()
+
+VERSIONS_TO_REGENERATE = (
```
Done

merwok commented on pep_sphinx_extensions/pep_zero_generator/pep_index_generator.py:
```diff
@@ -19,13 +19,20 @@
+if sys.version_info >= (3, 11):
+    from release_engineering.generate_release_cycle import create_release_cycle
+else:
+    # this function uses tomllib, which requires Python 3.11+
```
The comment would make more sense before line 31 or before line 30. _This_ function here on lines 34-35 does not use tomllib!

But also, you could be a version-dependent dependency like `tomli >= 1.1.0 ; python_version < "3.11"` and a try/except import. But it’s up to you, I’m not well versed in this repo to know whether RMs would use older pythons, and then whether a silent no-op is better than working or having a clear error 🙂

m-aciek reviewed (commented):
I've left some minor comments.

The centralised file is 3400 lines long. I wonder, would it be possible to get all release matadata from the release PEPs, after enforcing the data structure using Sphinx/docutils features? This way the source of truth would be PEPs with machine-extractable data, and the data would be much easier to maintain/add. Then JSON/TOML would be generated in CI or locally automatically from the PEPs. What do you think?

AA-Turner commented:
> I wonder, would it be possible to get all release matadata from the release PEPs, after enforcing the data structure using Sphinx/docutils features?

We could, but this would have a couple of considerations. Most importantly, we would need to change every previous release PEP, and this wouldn't easily permit going back before 1.6 if we wanted to record the metadata for 0.9-1.5.

I wouldn't be entirely opposed to this, though, as it preserves the nice quality that the PEPs remain the authoritative source for release information.

A

hugovk reviewed (commented):
Summary of discussion with Adam at EuroPython sprints:

I'd like us to provide the full data as JSON, so it's more widely available, for example, when using JavaScript.

I don't mind the source data being in TOML or JSON. It is extra work to convert from TOML to JSON, but that shouldn't be too hard, and TOML is a bit nicer to edit by hand.

I also don't mind if we provide the same smaller `release-cycle.json` file for the devguide to consume, or update the devguide to read the full one. Adam has preference for the small file.

AA-Turner commented on pep_sphinx_extensions/pep_zero_generator/pep_index_generator.py:
```diff
@@ -73,3 +74,9 @@ def create_pep_zero(app: Sphinx, env: BuildEnvironment, docnames: list[str]) ->
 
     write_peps_json(peps, Path(app.outdir))
+
+    release_cycle = create_release_cycle()
+    app.outdir.joinpath('release-cycle.json').write_text(release_cycle, encoding="utf-8")
+
+    release_json = create_release_json()
+    app.outdir.joinpath('python-releases.json').write_text(release_json, encoding="utf-8")
```
I don't think this is right. The releases as machine-readable data have nothing to do with the "API" of the PEPs website, to the extent that such a thing exists. I don't see a reasonable reason not to have the files at the root, it won't conflict with anything & it's a shorter link. If you do want to put the JSON files in a subdirectory though, let's use a more appropriate folder name (e.g. `/python-releases/{...}.json`).

hugovk replied:
> The releases as machine-readable data have nothing to do with the "API" of the PEPs website, to the extent that such a thing exists.

The data is here, it's coming from the PEPs website; it's data from the release PEPs.

We're precomputing the data to create a read-only API served via static JSON. This is a common thing to do.

> I don't see a reasonable reason not to have the files at the root

So we can document them via https://peps.python.org/api/

> it's a shorter link

This data will be mostly read by machines, an extra 4 characters is no problem.

AA-Turner commented:
Thanks for the review, I've responded to the comments (& fixed the formatting where it was wrong, thank you for pointing it out). I've also added the 1.6 alphas and corrected a few other dates/added more supporting links. I've updated the release manager for 1.6--2.1 to by Hilton & GvR, it seems that the current RM model didn't properly come into being until 2.1.

A

hugovk commented on release_management/python-releases.toml:
```diff
@@ -0,0 +1,3501 @@
+pep = 160
+status = "end-of-life"
+branch = ""  # no branch or tag for 1.6 exists
+release-manager = "Guido van Rossum & Jeremy Hylton"
```
CSV, like we have for 2.3 (3 RMs) and 3.3 (2 RMs)?
```suggestion
release-manager = "Guido van Rossum, Jeremy Hylton"
```

hugovk replied:
This is a format for machines to consume, would it be easier to deal with CSV than write a custom parser needing to handle "and" and ampersands?

(But I don't see a pressing need for splitting this list, so not a blocker.)

hugovk commented on release_management/python-releases.toml:
```diff
@@ -0,0 +1,3501 @@
+pep = 226
+status = "end-of-life"
+branch = "2.1"
+release-manager = "Guido van Rossum & Jeremy Hylton"
```
```suggestion
release-manager = "Guido van Rossum, Jeremy Hylton"
```

hugovk commented on release_management/python-releases.toml:
```diff
@@ -0,0 +1,3501 @@
+pep = 200
+status = "end-of-life"
+branch = "2.0"
+release-manager = "Guido van Rossum & Jeremy Hylton"
```
```suggestion
release-manager = "Guido van Rossum, Jeremy Hylton"
```

hugovk commented on release_management/__init__.py:
```diff
@@ -0,0 +1,81 @@
+    releases: dict[str, list[ReleaseInfo]]
+
+
+@dataclass(frozen=True,  **dc_kw)
```
I guess that's a no to Black? If not Black, then Ruff?

AA-Turner replied:
Your wish is my command (in this instance!) -- I've added Ruff formatting, the diff is minimal.

hugovk replied:
Thank you! 🎉

AA-Turner commented:
Python 3.9 failures are new and still related to the `pre-commit-uv` problems: https://github.com/tox-dev/pre-commit-uv/issues/70

A

ambv commented:
Just like with #4692, Imma be a tie breaker here and side with Hugo. I do think it's an API, even if it's read-only. But even if you disagree, this is such a minor detail, it's not worth blocking this otherwise great change to the release schedule PEP workflow.
